### DISCUSSION TITLE
Problem with yaw angle convergence

### DISCUSSION THREAD

**Discussion by na851998** [rodralez/NaveGo#138] [Q&A] [answered]
Hello,
Currently, I'm trying to use NaveGo with raw data. The problem is, when I initialize roll and pitch with random values, they can converge to the reference values, but it is not the case with yaw angle. So I was wondering, what may cause the problem.
(The reference yaw angle is ranged from 0 to 360 degrees so I subtract it by 180 degrees)
![untitled 1](https://user-images.githubusercontent.com/63492554/179960861-957b0232-b777-4192-bd8f-0c13669923cd.jpg)
![untitled1 1](https://user-images.githubusercontent.com/63492554/179960956-9091d0d0-8874-48d3-901d-0b2eca5c6858.jpg)
![untitled2 1](https://user-images.githubusercontent.com/63492554/179961021-d2e31659-6e18-4e37-afd1-a4090079ee60.jpg)

**Comment by rodralez** [comment, marked as answer]
A dead-reckoning systems, as an inertial navigation system, needs to known correct initial position, velocity and attitude (PVA). If you don't provide the correct PVA, PVA estimates will diverge.

**Comment by na851998** [comment]
Hi,
In real_ekinox example, I initilize roll, pitch and yaw as 2 rads, which are far different from true initial angles.
![fig1 1](https://user-images.githubusercontent.com/63492554/184108042-e0bfc62a-3d27-4df0-8b3c-b33079003064.jpg)
![fig2 1](https://user-images.githubusercontent.com/63492554/184108085-286003ec-6025-48ee-8287-92d913fd2c8c.jpg)
![fig3 1](https://user-images.githubusercontent.com/63492554/184108118-3811f0bc-1beb-47f3-8a43-5a4313a29d60.jpg)
As you can see, all angles converge to their true values. But when I use different dataset, only pitch and roll angles converge. 
I think there are two possibilities:
1) The dataset is wrong (but somehow pitch and roll still converge)
2) The conversion of yaw angle from range [0,2pi] to [-pi,pi] is wrong (yaw angle is subtracted by 2pi when value is greater than pi)

Thank you and look forward to your reply.
Kind regards.

**Reply by rodralez** [reply]
Hello,

You can try the following code to see how the yaw angle is corrected when is between 0 a 2pi.

```
clc
close all
clear
matlabrc

addpath ./ins/
addpath ./ins-gnss/
addpath ./ins-visual/
addpath ./conversions/

yaw_up = linspace(0, 2*pi, 300)';
yaw_down = linspace(2*pi, 0, 300)';
yaw = [yaw_up; yaw_down];

yaw_corr = correct_yaw(yaw);

figure
plot(yaw, '-', 'Color', 'green', 'LineWidth', 2)
hold on
plot(yaw_corr, '--', 'LineWidth', 2)
grid on
legend('yaw, from 0 to 2pi', 'yaw correct, from -pi to pi')
```

![image](https://user-images.githubusercontent.com/5432060/184991673-36c640e7-dfa3-433b-a942-711661236295.png)
